### PR TITLE
[404 server] Add prometheus metrics

### PR DESCRIPTION
**Pull request by cmluciano**
Metrics tracked:
- total number of requests
- duration of request
- status codes and method

I am representing duration for the request in milliseconds because my
test request almost always returned way under 1 second.

### CODE REVIEW

**Comment by cmluciano** [comment]
I've tried adding the prometheus library using godep, but the workspace directory keeps overriding it. I tried removing it since it will be unsupported in 1.8, but the build still fails while trying to locate prometheus.

**Comment by luxas** [comment]
cc @bprashanth

**Comment by cmluciano** [comment]
Any advice on the Godep pull @bprashanth ?

**Comment by piosz** [comment]
Try do the following:
```
rm -rf Godeps/_workspace
go get github.com/prometheus/client_golang/prometheus
go get github.com/prometheus/client_golang/prometheus/promhttp
godep save ./...
```

**Comment by cmluciano** [comment]
Not sure if it related to the AWS stuff, but I cannot retrieve the build log for some reason

**Comment by cmluciano** [comment]
Can i get a retest on this @piosz ? The Travis build seems to have exited prematurely.

**Comment by piosz** [comment]
Closed/reopened in order to retrigger tests.

**Comment by piosz** [comment]
Also you need to include `vendor` directory into your PR.

**Comment by cmluciano** [comment]
Fingers crossed :)

File: 404-server/Godeps/Godeps.json [hunk 1]
```diff
@@ -1,8 +1,15 @@
 {
 	"ImportPath": "k8s.io/contrib/404-server",
-	"GoVersion": "go1.5",
+	"GoVersion": "go1.7",
+	"GodepVersion": "v78",
 	"Packages": [
-		"./..."
+		"./.."
```
Comment (piosz) [inline 404-server/Godeps/Godeps.json]: You've made a mistake here. You should type:
```
godep save ./...
```
but you've typed
```
godep save ./..
```
The different is just one dot. The should be 3. Something weird had happened. Please try again.

**Comment by cmluciano** [comment]
Green Go the Rushes O !

Nice catch on the missing . @piosz

**Comment by piosz** [comment]
LGTM from me
@bowei can you please take a look or delegate?
@mikedanese can you please approve?

**Comment by piosz** [comment]
@cmluciano thanks for the PR

File: 404-server/server.go [hunk 2]
```diff
@@ -36,4 +37,10 @@ func main() {
 		fmt.Fprint(w, "ok")
 	})
 	http.ListenAndServe(fmt.Sprintf(":%d", *port), nil)
+	// TODO: Use .Shutdown from Go 1.8
+	err := http.ListenAndServe(fmt.Sprintf(":%d", *port), nil)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "could not start http server: %s\n", err)
```
Comment (bowei) [inline 404-server/server.go]: log.Fatal?

File: 404-server/server.go [hunk 3]
```diff
@@ -22,21 +22,45 @@ import (
...
+
 func main() {
 	flag.Parse()
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
 		w.WriteHeader(http.StatusNotFound)
 		fmt.Fprint(w, "default backend - 404")
+
+		duration := time.Now().Sub(start).Seconds() * 1e3
+		fmt.Println(duration)
```
Comment (bowei) [inline 404-server/server.go]: did you mean to print this to stdout?

File: 404-server/metrics.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		Name:      "request_count_total",
+		Help:      "Counter of HTTP requests made.",
+	}, []string{"proto"})
+
+	requestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "request_duration_milliseconds",
+		Help:      "Histogram of the time (in milliseconds) each request took.",
+		Buckets:   append([]float64{.001, .003}, prometheus.DefBuckets...),
```
Comment (bowei) [inline 404-server/metrics.go]: [minor] wonder if recording at microsecond resolution is useful?

File: 404-server/metrics.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	}, []string{"proto"})
+
+	responseStatus = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: namespace,
+		Subsystem: subsystem,
+		Name:      "response_status_count_total",
+		Help:      "Counter of response status codes.",
```
Comment (bowei) [inline 404-server/metrics.go]: isn't this always 404?

**Review by bowei — COMMENTED** [review]
some nits

File: 404-server/metrics.go (hunk 4 above)
Comment (cmluciano) [inline 404-server/metrics.go]: I did have this at microseconds, but I didn't know what the average response time would be "in the wild". Would you like me to switch to microseconds?

File: 404-server/server.go (hunk 3 above)
Comment (cmluciano) [inline 404-server/server.go]: Oops, that was for testing the time units, removing

**Comment by cmluciano** [comment]
I removed the request code part since it should always be a 404. I also removed the testing println that I missed. 

@bowei LMK what you think of https://github.com/kubernetes/contrib/pull/2376#discussion_r104457237

**Pull request merged**